### PR TITLE
Improving AddOpen code fix provider

### LINKED ISSUE #10521: visualstudio open namespace codefix has wrong indentation in some cases
In indented context, the open namespace code fix puts the declaration at wrong indentation level.

It doesn't always affect the code, but when there are type extensions, it breaks it.

```fsharp
module Foo =
  open System

  type DateTime with
    member x.X() = ()

  let a () =
    let t = Task.Factory.StartNew(fun () -> 1) // use the code fix on Task
    ()
```

open namespace indentiation glitch

**Related information**

VS2019 16.8

### LINKED ISSUE #15552: AddOpen code fix might add too many empty lines
Take a look:

https://github.com/dotnet/fsharp/assets/5451366/32d51035-e3cb-49bc-902e-928eed1291be

In the first case, the fix adds an extra line and then the open above it - which is reasonable.
In the second case, the fix adds two extra lines and then the open above it - which is okay but why.

This probably isn't a "bug" really, since the code compiles, but not an improvement either, since addressing this wouldn't bring new functionality. Just a small flaw kind of.

### LINKED ISSUE #15553: AddOpen code fix doesn't fix FS0043 anymore
Repro:

https://github.com/dotnet/fsharp/assets/5451366/ddb423f8-e63e-42e4-b8a1-1ed9308a23da

Code:

```fsharp
module M =
    let (++) x y = 10 * x + y

module N = 
    let theAnswer = 4 ++ 2
```

AFAIU this is supposed to work: https://github.com/dotnet/fsharp/pull/2474
Currently this should be triggered by FS0043 but the fix doesn't pop up.

### PR DESCRIPTION
**Pull request by psfinaki**
This addresses some, but not all, problems of the AddOpen code fix.

In particular:
- fixes #15552
- fixes #15553
- partially fixes #10521 - the resulting code is not broken anymore but the insertion point is not exact yet

The code is crap but I am keeping most of it for now no minimize the risks.

### CODE REVIEW

File: vsintegration/src/FSharp.Editor/CodeFixes/AddOpenCodeFixProvider.fs [hunk 1]
```diff
@@ -50,128 +111,125 @@ type internal AddOpenCodeFixProvider [<ImportingConstructor>] (assemblyContentPr
...
+                                    parseResults.ParseTree
+                                    maybeUnresolvedIdents
+                                    insertionPoint
+
+                            entities
+                            |> Seq.map createEntity
+                            |> Seq.concat
+                            |> Seq.toList
+                            |> getSuggestionsAsCodeFixes context sourceText
+                            |> Seq.tryHead))
```
Comment (psfinaki) [inline vsintegration/src/FSharp.Editor/CodeFixes/AddOpenCodeFixProvider.fs:232]: Same case as in #15891
The code fix already offers only the first thing, hence also #15911

Will be fixed separately.

**Review by T-Gro — APPROVED** [review]

File: vsintegration/src/FSharp.Editor/CodeFixes/AddOpenCodeFixProvider.fs [hunk 2]
```diff
@@ -2,40 +2,101 @@
...
+                // nested module, shouldn't be here
+                | line when line.StartsWith "module" -> insertionLineNumber, $"{margin}open {ns}{br}{br}"
+
+                // attribute, shouldn't be here
+                | line when line.StartsWith "[<" && line.EndsWith ">]" ->
+                    let moduleDeclLineNumber =
+                        sourceText.Lines
+                        |> Seq.skip insertionLineNumber
+                        |> Seq.findIndex (fun line -> line.ToString().Contains "module")
+                        |> (+) insertionLineNumber
```
Comment (vzarytovskii) [inline vsintegration/src/FSharp.Editor/CodeFixes/AddOpenCodeFixProvider.fs:62]: can this please be a normal lambda? Partially applied operators are hard to read, even if it's as simple as addition. We shouldn't really encourage it.

Comment (psfinaki) [reply vsintegration/src/FSharp.Editor/CodeFixes/AddOpenCodeFixProvider.fs]: Sure, there you go!

**Review by 0101 — APPROVED** [review]
Solid testing!

**Pull request merged**